### DISCUSSION TITLE
Backporting Clustering Indices to SciKit-Learn

### DISCUSSION THREAD

**Discussion by BradKML** [online-ml/river#778] [Q&A] [answered]
Currently I have asked this question for https://github.com/scikit-learn/scikit-learn/issues/11369#issuecomment-972591299 and https://github.com/DistrictDataLabs/yellowbrick/issues/1205 that there are some indices that are not in the major SciKit repos, and I see that your repo includes some of the included indices. However Online-ML is a specialized form of Machine Learning that is not necessarily useful for standard clustering, would it be possible to request a backport @hoanganhngo610 ?
https://github.com/online-ml/river/tree/main/river/metrics/cluster

**Comment by MaxHalford** [comment, marked as answer]
Hello there 👋 

I'm sorry but I don't really understand your message. What are you suggesting exactly? What do you want to backport?

**Reply by BradKML** [reply]
Currently, there is a package called YellowBrick that find the optimal amounts of clusters for a multi-dimensional distribution.
https://github.com/DistrictDataLabs/yellowbrick/issues/1205
But they rely on SciKit-Learn as a dependency, which some of the indices in River have but not in SciKit-Learn
https://github.com/scikit-learn/scikit-learn/issues/11369#issuecomment-972591299
Is it possible to add the cluster metrics to SciKit-Learn (Dunn’s, Xie-Beni, SD validity, S_Dbw among others)?
Previously noted in https://github.com/scikit-learn/scikit-learn/issues/11369#issuecomment-400962824

**Reply by MaxHalford** [reply]
>  Is it possible to add the cluster metrics to SciKit-Learn (Dunn’s, Xie-Beni, SD validity, S_Dbw among others)?

I imagine it is. But we're not going to do that for you! :)

**Reply by BradKML** [reply]
That is fair. Will ask about some of the internal object methods later in time.
(init => update => get, as one streamline function)

**Reply by BradKML** [reply]
@MaxHalford question: why do the functions of the evaluation algorithms count the number of points? Does the update function update by the point rather by aggregate?

**Reply by MaxHalford** [reply]
I'm really sorry but I don't understand anything you're saying 😅. For instance, "update by the point" doesn't mean anything to me.

**Reply by BradKML** [reply]
@MaxHalford the Cluster Accuracy Metrics updates with `x, y_pred, centers, sample_weight=1.0`, which in the case of SciKit-Learn and others, are often Numpy Arrays of "number of points" X "dimensions of vectors" (with Labels being the other variable).
https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/metrics/cluster/_unsupervised.py

However, if that is the same case for River, the "total_points" would not make sense as the update method in the Cluster evaluation function.
- https://github.com/online-ml/river/blob/main/river/metrics/cluster/rmsstd.py#L67
- https://github.com/online-ml/river/blob/main/river/metrics/cluster/ssb.py#L71
- https://github.com/online-ml/river/blob/main/river/metrics/cluster/ssq_based.py#L73
- https://github.com/online-ml/river/blob/main/river/metrics/cluster/ssw.py#L254 
- https://github.com/online-ml/river/blob/main/river/metrics/cluster/xiebeni.py#L79

**Reply by hoanganhngo610** [reply]
@BrandonKMLee As you may know, in online ML, we do not get to keep the information of all the points (and their labels) that have been passed through previously. This leads to the principle difference in the implementation of metrics of `River` and `Scikit-learn`, and as a work-around, we have to implement a counter in order to count the total number of points passed through in the data stream.

**Reply by BradKML** [reply]
Sorry for asking, but the folder for these indices are seemingly gone? I may have missed something. https://github.com/online-ml/river/pull/484

**Reply by smastelini** [reply]
Hi @BrandonKMLee, these metrics are in the river-extra repo for the time being.
